src/Compilers/CSharp/Portable/Symbols/NativeIntegerTypeSymbol.cs

```diff
@@ -93,50 +172,239 @@ internal NativeIntegerTypeSymbol(NamedTypeSymbol underlyingType) : base(underlyi
+        /// <summary>
+        /// Replaces references to underlying type with references to native integer type.
+        /// </summary>
+        internal TypeWithAnnotations SubstituteUnderlyingType(TypeWithAnnotations type) => type.SubstituteType(GetTypeMap());
+
+        /// <summary>
+        /// Replaces references to underlying type with references to native integer type.
+        /// </summary>
+        internal NamedTypeSymbol SubstituteUnderlyingType(NamedTypeSymbol type) => GetTypeMap().SubstituteNamedType(type);
+
+        [Conditional("DEBUG")]
+        internal static void VerifyEquality(Symbol nativeIntegerSymbol, Symbol underlyingSymbol)
+        {
+            Debug.Assert(!nativeIntegerSymbol.Equals(underlyingSymbol, TypeCompareKind.ConsiderEverything));
+            Debug.Assert(!underlyingSymbol.Equals(nativeIntegerSymbol, TypeCompareKind.ConsiderEverything));
+            Debug.Assert(nativeIntegerSymbol.Equals(underlyingSymbol, TypeCompareKind.IgnoreNativeIntegers));
+            Debug.Assert(underlyingSymbol.Equals(nativeIntegerSymbol, TypeCompareKind.IgnoreNativeIntegers));
+            Debug.Assert(nativeIntegerSymbol.GetHashCode() == underlyingSymbol.GetHashCode());
+        }
+
+        private sealed class NativeIntegerTypeMap : AbstractTypeMap
+        {
+            private readonly NativeIntegerTypeSymbol _type;
+            private readonly SpecialType _specialType;
+
+            internal NativeIntegerTypeMap(NativeIntegerTypeSymbol type)
+            {
+                _type = type;
+                _specialType = _type.UnderlyingNamedType.SpecialType;
+
+                Debug.Assert(_specialType == SpecialType.System_IntPtr || _specialType == SpecialType.System_UIntPtr);
+            }
+
+            internal override NamedTypeSymbol SubstituteTypeDeclaration(NamedTypeSymbol previous)
+            {
+                return previous.SpecialType == _specialType ?
+                    _type :
+                    base.SubstituteTypeDeclaration(previous);
+            }
+        }
+    }
+
+    internal sealed class NativeIntegerMethodSymbol : WrappedMethodSymbol, Cci.IReference
+    {
+        private readonly NativeIntegerTypeSymbol _container;
+        private readonly Symbol? _associatedSymbol;
+
+        internal NativeIntegerMethodSymbol(NativeIntegerTypeSymbol container, MethodSymbol underlyingMethod, Symbol? associatedSymbol)
+        {
+            Debug.Assert(!underlyingMethod.IsGenericMethod);
+            _container = container;
+            _associatedSymbol = associatedSymbol;
+            UnderlyingMethod = underlyingMethod;
+            NativeIntegerTypeSymbol.VerifyEquality(this, underlyingMethod);
+        }
+
+        public override Symbol ContainingSymbol => _container;
+
+        public override MethodSymbol UnderlyingMethod { get; }
+
+        public override TypeWithAnnotations ReturnTypeWithAnnotations => _container.SubstituteUnderlyingType(UnderlyingMethod.ReturnTypeWithAnnotations);
```

Comment:
>_container.SubstituteUnderlyingType(UnderlyingMethod.ReturnTypeWithAnnotations) [](start = 73, length = 79)

Should we go through the trouble of substituting custom modifiers? It looks like we don't substitute RefCustomModifiers. Feels inconsistent.

Reply:
Note that `TypeWithAnnotations.Equals()` currently ignores `TypeCompareKind` when comparing custom modifiers.

---
In reply to: [418340471](https://github.com/dotnet/roslyn/pull/43766#discussion_r418340471) [](ancestors = 418340471)

Reply:
> TypeWithAnnotations.Equals() currently ignores TypeCompareKind when comparing custom modifiers

This is expected behavior since annotations are not supposed to affect custom modifiers

---
In reply to: [419022649](https://github.com/dotnet/roslyn/pull/43766#discussion_r419022649) [](ancestors = 419022649,418340471)